src/storage/src/read/windowed.rs

```diff
@@ -0,0 +1,137 @@
+            batches.push(
+                batch
+                    .columns
+                    .into_iter()
+                    .map(|v| v.to_arrow_array())
+                    .collect::<Vec<_>>(),
+            );
+        }
+
+        let num_columns = batches.get(0).map(|b| b.len()).unwrap_or(0);
+        let mut vectors_in_batch = Vec::with_capacity(num_columns);
+
+        for idx in 0..num_columns {
+            let columns: Vec<&dyn Array> =
+                batches.iter().map(|b| b[idx].as_ref()).collect::<Vec<_>>();
+            vectors_in_batch
+                .push(arrow::compute::concat(&columns).context(error::ConvertColumnsToRowsSnafu)?);
+        }
+        let sorted = sort_by_rows(&self.schema, vectors_in_batch)?;
+        let vectors = sorted
+            .iter()
+            .zip(store_schema.columns().iter().map(|c| &c.desc.name))
+            .map(|(arr, name)| {
+                Helper::try_into_vector(arr).context(error::ConvertChunkSnafu { name })
+            })
+            .collect::<Result<_>>()?;
+        Ok(Some(Batch::new(vectors)))
+    }
+}
+
+fn sort_by_rows(schema: &ProjectedSchemaRef, arrays: Vec<ArrayRef>) -> Result<Vec<ArrayRef>> {
+    let sort_columns = build_sort_columns(schema);
+    let store_schema = schema.schema_to_read();
+    // Convert columns to rows to speed lexicographic sort
+    let mut row_converter = RowConverter::new(
+        sort_columns
+            .iter()
+            .map(|(idx, descending)| {
+                SortField::new_with_options(
+                    store_schema.columns()[*idx].desc.data_type.as_arrow_type(),
+                    SortOptions {
+                        descending: *descending,
+                        nulls_first: true,
+                    },
+                )
+            })
+            .collect(),
+    )
+    .context(error::ConvertColumnsToRowsSnafu)?;
+
+    let columns_to_sort = sort_columns
+        .into_iter()
+        .map(|(idx, _)| arrays[idx].clone())
+        .collect::<Vec<_>>();
+
+    let rows_to_sort = row_converter
+        .convert_columns(&columns_to_sort)
+        .context(error::ConvertColumnsToRowsSnafu)?;
+
+    let mut sort_pairs = rows_to_sort.iter().enumerate().collect::<Vec<_>>();
+    sort_pairs.sort_unstable_by(|(_, a), (_, b)| a.cmp(b));
+
+    let idx =
+        arrow::array::UInt32Array::from_iter_values(sort_pairs.iter().map(|(i, _)| *i as u32));
+
+    let sorted = arrays
+        .iter()
+        .map(|arr| arrow::compute::take(arr, &idx, None))
+        .collect::<arrow::error::Result<Vec<_>>>()
+        .context(error::SortArraysSnafu)?;
+
+    debug_assert_eq!(sorted.len(), store_schema.num_columns());
+
+    Ok(sorted)
+}
+
+/// [<PK_1>, <PK_2>, TS] to [TS, <PK_1>, <PK_2>]
+fn build_sort_columns(schema: &ProjectedSchemaRef) -> Vec<(usize, bool)> {
+    let row_key_end = schema.schema_to_read().row_key_end();
+    let ts_col_index = row_key_end - 1;
```

Comment:
I think we can add a function `time_column_index` for `StoreSchema` to get the time column index. We should not hard code the implicit time index position of row keys in other places. It's not maintainable.